### PR TITLE
Consumer timeout for transfer processes

### PR DESCRIPTION
**Pull request by marcgs**
An extension providing watchdog process (TransferProcessWatchdog) that monitors long running transfers, and automatically move these into the ERROR state after a configurable timeout. 

A patch for the TransferProcessManagerImpl has been implemented until the PR solving issue [#324](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/324) is merged.

### CODE REVIEW

File: coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        }
+    }
+
+    private void run() {
+        List<TransferProcess> transferProcesses = transferProcessStore.nextForState(IN_PROGRESS.code(), batchSize);
+        transferProcesses.stream()
+            .filter(p -> p.getStateTimestamp() > stateTimeout)
+            .forEach(p -> {
+                p.transitionError("Timeout");
+                transferProcessStore.update(p);
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java:64]: manage optimistic concurrency exceptions?

File: coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+    private TransferProcessStore transferProcessStore;
+    private ScheduledExecutorService executor;
+
+    @Builder
+    private TransferProcessWatchdog(Monitor monitor, int batchSize, long delayInSeconds, long stateTimeout) {
+        this.monitor = monitor;
+        this.batchSize = batchSize;
+        this.delayInSeconds = delayInSeconds;
+        this.stateTimeout = stateTimeout;
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java:43]: wouldn't the timeout value make more sense as an (optional) attribute of the transfer itself?

File: coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/pom.xml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    <groupId>net.catenax.prs</groupId>
+    <artifactId>edc-transfer-process-watchdog</artifactId>
+
+    <properties>
+        <checkstyle.skip>true</checkstyle.skip>
+        <pmd.skip>true</pmd.skip>
+    </properties>
+
+    <name>EDC Transfer Process Watchdog</name>
+    <description>Module to clean up hanging transfer processes</description>
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/pom.xml:21]: somewhat vague

Comment (marcgs) [reply coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/pom.xml:21]: There is a more thorough description in the README. I think we should keep these entries in the pom.xml short. But I am open for suggestions for rewordings :-)

File: coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java (hunk 1 above)
Comment (marcgs) [reply coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java:64]: There is unfortunately not a standardised way of handling these kind of exceptions with the `TransferProcessStore`. At the moment, concurrency management is completely handled by the specific implementation (Azure, AWS, in memory...) and not exposed in any way in the SPI interfaces.

File: coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java (hunk 2 above)
Comment (marcgs) [reply coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java:43]: I think we could have both the possibility of specifying a timeout specific for a transfer as well as a global one.

Comment (marcgs) [reply coreservices/partsrelationshipservice/connector/edc-transfer-process-watchdog/src/main/java/org/eclipse/dataspaceconnector/transfer/core/transfer/TransferProcessWatchdog.java:43]: I would make this change in a separate PR though. One thing to consider is that we need the patched core in order to be able to pass properties in the `TransferProcess`. At the moment the change does not depend on this and could be contributed to EDC as is.

**Pull request closed without merging**